### PR TITLE
feat: update bindgen for dojo.js sdk

### PR DESCRIPTION
**Pull request by MartianGreed**
# Description

**this is a draft** I've left out code comments in code that will be removed afterward.

Revamping bindgen to make it more modular and more extandable.

We can discuss on `BindgenGenerator` which may evolve.

I think this could be a great base to work. After finishing update for `--typescript` I will add a `--recs` to add retro compatibility.

## Tests

- [x] Yes
- [ ] No, because they aren't needed
- [ ] No, because I need help

## Added to documentation?

In code itself, as this is a very specific part of software. Might add a README to guide developper through creation of new plugins like godot for instance

- [ ] README.md
- [ ] [Dojo Book](https://github.com/dojoengine/book)
- [ ] No documentation needed

## Checklist

- [ ] I've formatted my code (`scripts/prettier.sh`, `scripts/rust_fmt.sh`, `scripts/cairo_fmt.sh`)
- [ ] I've linted my code (`scripts/clippy.sh`, `scripts/docs.sh`)
- [x] I've commented my code
- [ ] I've requested a review after addressing the comments

## Summary by CodeRabbit

- **New Features**
	- Introduced `TypescriptRecsPlugin` for enhanced TypeScript code generation from Dojo models and contracts.
	- Added support for generating TypeScript enums, interfaces, functions, and schemas through new generator structures.
	- Modularized code generation process with multiple writers for improved extensibility.
	- Added a new dependency for logging functionality.

- **Bug Fixes**
	- Updated tests to ensure compatibility with the new `TypescriptRecsPlugin`.

- **Documentation**
	- Added comprehensive test modules for new generator functionalities.

- **Refactor**
	- Improved structure of `TypescriptPlugin` for better code generation handling and maintainability.

### CODE REVIEW

**Comment by rsodre** [comment]
Generated components should also include the new Torii indexed ERC20/ERC721 models, so they can be synched by the  Torii client, implemented on #2494 

These are the models, from graphQL schema:
```graphql
type ERC__Balance {
  balance: String!
  type: String!
  tokenMetadata: ERC__Token!
}

type ERC__Token {
  name: String!
  symbol: String!
  tokenId: String!
  decimals: String!
  contractAddress: String!
}

type ERC__Transfer {
  from: String!
  to: String!
  amount: String!
  type: String!
  executedAt: String!
  tokenMetadata: ERC__Token!
  transactionHash: String!
}
```

**Comment by rsodre** [comment]
@MartianGreed, generated components should also include the new Torii indexed ERC20/ERC721 models, so they can be synched by the  Torii client, implemented on #2494 

These are the models, from graphQL schema:
```graphql
type ERC__Balance {
  balance: String!
  type: String!
  tokenMetadata: ERC__Token!
}

type ERC__Token {
  name: String!
  symbol: String!
  tokenId: String!
  decimals: String!
  contractAddress: String!
}

type ERC__Transfer {
  from: String!
  to: String!
  amount: String!
  type: String!
  executedAt: String!
  tokenMetadata: ERC__Token!
  transactionHash: String!
}
```

**Comment by MartianGreed** [comment]
@rsodre do you want to query those models over grpc or over graphql ? Or maybe both ?

**Comment by rsodre** [comment]
those are already available on GraphQL, example [here](https://github.com/rsodre/dojo-oz-token/blob/main/client/src/TokensQL.tsx).
I'm asking to have them sync with grpc too.

**Comment by glihm** [comment]
@MartianGreed will get back to it shortly and let a review. Thanks for your patience. 🙏

File: crates/dojo-bindgen/src/plugins/recs/mod.rs [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+        {
+            return String::new();
+        }
+        fn map_type(token: &Token) -> String {
+            match token {
+                Token::CoreBasic(_) => TypescriptRecsPlugin::map_type(token)
+                .replace("RecsType.", "").replace("Array", "[]")
+                // types should be lowercased
+                .to_lowercase(),
+                Token::Composite(t) => format!("models.{}", t.type_name()),
```
Comment (glihm) [inline crates/dojo-bindgen/src/plugins/recs/mod.rs:359]: Are you sure at this level the composites tokens are only models?
You may end up with some enums that are simply types passed as inputs to systems?

Could be interesting to test with some enums and structs the are not models. Or models is a generic module where you have all the types?

File: crates/dojo-bindgen/src/plugins/recs/mod.rs [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+            out += TypescriptRecsPlugin::format_model(&namespace, &model).as_str();
+        }
+
+        out += "    };
+}\n";
+
+        out
+    }
+
+    // Formats a system into a C# method used by the contract class
```
Comment (glihm) [inline crates/dojo-bindgen/src/plugins/recs/mod.rs:336]:
```suggestion
    // Formats a system into a Typescript method used by the contract class
```
In a next pass, could be helpful having an adjustment of comments to the very specific use-case here to contextualize more the reader.

File: crates/dojo-bindgen/src/plugins/typescript/generator/enum.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        let token_dup = create_available_theme_enum_token();
+        let result = writer.generate(&token_dup, &mut buff).unwrap();
+        assert_eq!(buff.len(), 1);
+        assert!(result.is_empty())
+    }
+
+    fn create_available_theme_enum_token() -> Composite {
+        Composite {
+            type_path: "core::test::AvailableTheme".to_owned(),
```
Comment (glihm) [inline crates/dojo-bindgen/src/plugins/typescript/generator/enum.rs:111]:
```
            type_path: "core::test::AvailableTheme".to_owned(),
```
Those type of values that we have several time may be extracted in constants to ease changes.

**Review by glihm — APPROVED** [review]
Amazing for the first iteration on that @MartianGreed!

Some minor comments that will be good for subsequent PRs.
Also, for all the indentation with `\t`, spaces may be more common. Also, maybe we could generate something without caring that much about format, and part of integrating generated code, the commit may also include custom formatting.

It would also ease the reading in the code itself (in the rust code).

Let's merge and iterate!

**Pull request merged**